pkg/pillar/evetpm/tpm.go

```diff
@@ -0,0 +1,189 @@
+// Copyright (c) 2020 Zededa, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package evetpm
+
+import (
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/x509"
+	"encoding/asn1"
+	"encoding/pem"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"math/big"
+	"os"
+	"unsafe"
+
+	"github.com/google/go-tpm/tpm2"
+	"github.com/google/go-tpm/tpmutil"
+	"github.com/lf-edge/eve/pkg/pillar/types"
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	//TpmDevicePath is the TPM device file path
+	TpmDevicePath = "/dev/tpmrm0"
+
+	//TpmDeviceKeyHdl is the well known TPM permanent handle for device key
+	TpmDeviceKeyHdl tpmutil.Handle = 0x817FFFFF
+
+	//TpmPasswdHdl is the well known TPM NVIndex for TPM Credentials
+	TpmPasswdHdl tpmutil.Handle = 0x1600000
+
+	//TpmCredentialsFileName is the file that holds the dynamically created TPM credentials
+	TpmCredentialsFileName = types.IdentityDirname + "/tpm_credential"
+
+	//MaxPasswdLength is the max length allowed for a TPM password
+	MaxPasswdLength = 7 //limit TPM password to this length
+)
+
+//TpmPrivateKey is Custom implementation of crypto.PrivateKey interface
+type TpmPrivateKey struct {
+	PublicKey crypto.PublicKey
+}
+
+//Helper structure to pack ecdsa signature for ASN1 encoding
+type ecdsaSignature struct {
+	R, S *big.Int
+}
+
+//Public implements crypto.PrivateKey interface
+func (s TpmPrivateKey) Public() crypto.PublicKey {
+	clientCertBytes, err := ioutil.ReadFile(types.DeviceCertName)
+	if err != nil {
+		return nil
+	}
+	block, _ := pem.Decode(clientCertBytes)
+	var cert *x509.Certificate
+	cert, _ = x509.ParseCertificate(block.Bytes)
+	ecdsaPublicKey := cert.PublicKey.(*ecdsa.PublicKey)
+	return ecdsaPublicKey
+}
+
+//Sign implements cryto.PrivateKey interface
+func (s TpmPrivateKey) Sign(r io.Reader, digest []byte, opts crypto.SignerOpts) ([]byte, error) {
+	R, S, err := TpmSign(digest)
+	if err != nil {
+		return nil, err
+	}
+	return asn1.Marshal(ecdsaSignature{R, S})
+}
+
+//ReadOwnerCrdl returns credential specific to this device
+func ReadOwnerCrdl() (string, error) {
+	tpmOwnerPasswdBytes, err := ioutil.ReadFile(TpmCredentialsFileName)
+	if err != nil {
+		return "", err
+	}
+	tpmOwnerPasswd := string(tpmOwnerPasswdBytes)
+	if len(tpmOwnerPasswd) > MaxPasswdLength {
+		tpmOwnerPasswd = tpmOwnerPasswd[0:MaxPasswdLength]
+	}
+	return tpmOwnerPasswd, nil
+}
+
+//TpmSign is used by external packages to get a digest signed by
+//device key in TPM
+func TpmSign(digest []byte) (*big.Int, *big.Int, error) {
+
+	rw, err := tpm2.OpenTPM(TpmDevicePath)
+	if err != nil {
+		return nil, nil, err
+	}
+	defer rw.Close()
+
+	tpmOwnerPasswd, err := ReadOwnerCrdl()
+	if err != nil {
+		log.Fatalf("Error in fetching TPM credentials: %v", err)
```

Comment:
Here you fatal, but in all other cases where ReadOwnerCrdl fails you log.Error.

When would it fail? Should you check up front and disable the tpm if it fails (and then log fatal everywhere) or should you make this function return an error?

Reply:
It is indeed a fatal, since device key can't be used without knowing the credential. 
a) If we fail to create tpm_credential, we will switch to software based cert (done in generate-self-signed)
b) in case someone deletes /config/tpm_credential tpmmgr tries to restore it when it is started as a service.
Therefore changed other places to do Fatal if ReadOwnerCrdl() fails.

Reply:
OK